File: ui-ios/Sources/ThreadwireUI/Theme/ThreadwireTypography.swift
import SwiftUI
import SwiftStreamingMarkdown

/// The design's 3-tier font scale - `.medium` is this milestone's hardcoded default
/// (see ThreadwireColors' ChatUIConfig note).
public enum ThreadwireFontScale {
    case small, medium, large
}

/// A Markdown render config aligned to Threadwire's design tokens. `MarkdownRenderConfig`'s
/// `.default` renders body text with the library's bundled 17pt `baseTextFonts` in *its*
/// theme color - both larger than the design's message size (medium `msg` = 15.5) and not
/// our text token, which is why message text looked oversized on iOS versus the design and
/// versus Android. This overrides the paragraph/list body to the design's size + the
/// caller's color (white in the user bubble, `colors.text` in the assistant bubble).
/// Headings/tables/code keep the library defaults on purpose (larger / monospaced).
func threadwireMarkdownConfig(textColor: Color, bodySize: CGFloat = 15.5) -> MarkdownRenderConfig {
    let fonts = TextFonts(
        normal: .systemFont(ofSize: bodySize, weight: .regular),
        italic: .italicSystemFont(ofSize: bodySize),
        bold: .systemFont(ofSize: bodySize, weight: .semibold),
        boldItalic: nil,
        preferredLetterSpacing: nil,
        preferredLineHeight: nil
    )
    let body = MarkdownRenderConfig.MarkdownTextStyle(textFonts: fonts, textColor: textColor)
    return MarkdownRenderConfig(orderedListStyle: body, paragraphStyle: body)
}

public struct ThreadwireTypography {
    public let title, header, msg, meta, name: Font
}

/// Sizes (pt) transcribed verbatim from the design's `FS` table - converted 1:1 from the
/// design's px values, the standard iOS convention.
func threadwireTypography(_ scale: ThreadwireFontScale) -> ThreadwireTypography {
    switch scale {
    case .small:
        return ThreadwireTypography(
            title: .system(size: 20, weight: .bold),
            header: .system(size: 15, weight: .semibold),
            msg: .system(size: 14),
            meta: .system(size: 12),
            name: .system(size: 15, weight: .medium)
        )
    case .medium:
        return ThreadwireTypography(
            title: .system(size: 23, weight: .bold),
            header: .system(size: 16, weight: .semibold),
            msg: .system(size: 15.5),
            meta: .system(size: 12.5),
            name: .system(size: 16, weight: .medium)
        )
    case .large:
        return ThreadwireTypography(
            title: .system(size: 26, weight: .bold),
            header: .system(size: 17, weight: .semibold),
            msg: .system(size: 17),
            meta: .system(size: 13.5),
            name: .system(size: 17, weight: .medium)
        )
    }
}

/// Card / generic container corner radius (design: card wrappers, chips use ~14-16).
let threadwireCardCornerRadius: CGFloat = 16

/// Message-bubble shapes, transcribed from the design's `bubbleUser` (`18 18 4 18`) and
/// `bubbleAssistant` (`18 18 18 4`) — asymmetric "tail" corner. CSS order is
/// TL/TR/BR/BL; `UnevenRoundedRectangle` (iOS 16.0+) takes topLeading/bottomLeading/
/// bottomTrailing/topTrailing.
let threadwireUserBubbleShape = UnevenRoundedRectangle(
    topLeadingRadius: 18, bottomLeadingRadius: 18, bottomTrailingRadius: 4, topTrailingRadius: 18
)
let threadwireAssistantBubbleShape = UnevenRoundedRectangle(
    topLeadingRadius: 18, bottomLeadingRadius: 4, bottomTrailingRadius: 18, topTrailingRadius: 18
)

File: ui-android/src/main/kotlin/com/fsk/threadwire/ui/theme/ThreadwireTypography.kt
package com.fsk.threadwire.ui.theme

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** The design's 3-tier font scale - `medium` is this milestone's hardcoded default (see ThreadwireColors' ChatUIConfig note). */
enum class ThreadwireFontScale { SMALL, MEDIUM, LARGE }

data class ThreadwireTypography(
    val title: TextStyle,
    val header: TextStyle,
    val msg: TextStyle,
    val meta: TextStyle,
    val name: TextStyle,
)

/** Sizes (sp) transcribed verbatim from the design's `FS` table. */
fun threadwireTypography(scale: ThreadwireFontScale): ThreadwireTypography = when (scale) {
    ThreadwireFontScale.SMALL -> ThreadwireTypography(
        title = TextStyle(fontSize = 20.sp, fontWeight = FontWeight.Bold),
        header = TextStyle(fontSize = 15.sp, fontWeight = FontWeight.SemiBold),
        msg = TextStyle(fontSize = 14.sp),
        meta = TextStyle(fontSize = 12.sp),
        name = TextStyle(fontSize = 15.sp, fontWeight = FontWeight.Medium),
    )
    ThreadwireFontScale.MEDIUM -> ThreadwireTypography(
        title = TextStyle(fontSize = 23.sp, fontWeight = FontWeight.Bold),
        header = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
        msg = TextStyle(fontSize = 15.5.sp),
        meta = TextStyle(fontSize = 12.5.sp),
        name = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Medium),
    )
    ThreadwireFontScale.LARGE -> ThreadwireTypography(
        title = TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold),
        header = TextStyle(fontSize = 17.sp, fontWeight = FontWeight.SemiBold),
        msg = TextStyle(fontSize = 17.sp),
        meta = TextStyle(fontSize = 13.5.sp),
        name = TextStyle(fontSize = 17.sp, fontWeight = FontWeight.Medium),
    )
}

/** Card / generic container corner radius (design: card wrappers, chips use ~14-16). */
const val ThreadwireCardCornerRadiusDp = 16

/**
 * Message-bubble corner radii, transcribed from the design's `bubbleUser`/`bubbleAssistant`
 * (`18px 18px 4px 18px` and `18px 18px 18px 4px`) — asymmetric "tail" corners: user's tail
 * is bottom-end (right), assistant's is bottom-start (left). CSS order is
 * top-left/top-right/bottom-right/bottom-left; Compose order is topStart/topEnd/bottomEnd/bottomStart.
 */
val ThreadwireUserBubbleShape = androidx.compose.foundation.shape.RoundedCornerShape(
    topStart = 18.dp, topEnd = 18.dp, bottomEnd = 4.dp, bottomStart = 18.dp,
)
val ThreadwireAssistantBubbleShape = androidx.compose.foundation.shape.RoundedCornerShape(
    topStart = 18.dp, topEnd = 18.dp, bottomEnd = 18.dp, bottomStart = 4.dp,
)
